firedrake/function.py

```diff
@@ -393,108 +383,34 @@ def assign(self, expr, subset=None):
         :class:`Function`'s ``node_set``.  The expression will then
         only be assigned to the nodes on that subset.
         """
-        # Avoid generating code when assigning scalar values to the Real space
-        if (isinstance(expr, numbers.Number)
-                and self.function_space().ufl_element().family() == "Real"):
-            self.dat.data[...] = expr
-            return self
-
-        expr = ufl.as_ufl(expr)
-        if isinstance(expr, ufl.classes.Zero):
-            self.dat.zero(subset=subset)
-            return self
-        elif (isinstance(expr, Function)
-              and expr.function_space() == self.function_space()):
-            expr.dat.copy(self.dat, subset=subset)
-            return self
-
-        from firedrake import assemble_expressions
-        assemble_expressions.evaluate_expression(
-            assemble_expressions.Assign(self, expr), subset)
+        from firedrake.assign import Assigner
```

Comment:
Is there a reason we can't `from firedrake import assign` at the top and `assign.Assigner.assign()`? Or just import the things we need from `assign.py`

Reply:
Because `firedrake/assign.py` has `from firedrake.function import Function` at the top (import loop).